cmd/juju/commands/switch.go

```diff
@@ -74,9 +91,34 @@ func (c *switchCommand) Info() *cmd.Info {
 	})
 }
 
+// SetFlags implements Command.SetFlags.
+func (c *switchCommand) SetFlags(f *gnuflag.FlagSet) {
+	c.CommandBase.SetFlags(f)
+	f.StringVar(&c.modelName, "m", "", "Model to operate in. Accepts [<controller name>:]<model name>")
+	f.StringVar(&c.modelName, "model", "", "")
+	f.StringVar(&c.controllerName, "c", "", "Controller to operate in")
+	f.StringVar(&c.controllerName, "controller", "", "")
+}
+
 func (c *switchCommand) Init(args []string) error {
 	var err error
-	c.Target, err = cmd.ZeroOrOneArgs(args)
+	var target string
+	if c.modelName != "" && c.controllerName != "" {
+		return errors.Trace(fmt.Errorf("cannot specify both a --model and --controller"))
+	}
+	if c.modelName != "" || c.controllerName != "" {
+		// translate `-c ctrl` to `ctrl:` and `-m model` to `:model`
+		target = fmt.Sprintf("%s:%s", c.controllerName, c.modelName)
+		err = cmd.CheckEmpty(args)
```

Comment:
This doesn't feel nice. Do we really need to join the model and controller name? Later down the line we split them, I think we should keep them separate here.

We could even move the (potential) split of c.Target into Init here

I also don't think we need to support `:-` and `-:` anymore (in fact I think `-:` would error now because it's not a flag?). Can we bypass this entirely?

Reply:
This is turning out to be very fiddly :sweat_smile:

Reply:
It's because amho, the `-` is more of a trick. Main code support `controller:`, `:model`, `controller-or-model` or `controller:model`. So it make sens to use those format as go-for-all. But I understand your feeling, even if I am not sure that we may do better. The fact that we need to handle `controller-or-model` usecase make separating controller and model a tricky case. But maybe we can merge both and have a function that retun `(controller, model string, err error)` and just set Controller & model in Run. however I am not sure that the code would be more readable.

Could need a second pair of eyes when doing that stuff... Are you open to peer prog on monday ?

Reply:
> It's because amho, the `-` is more of a trick. Main code support `controller:`, `:model`, `controller-or-model` or `controller:model`. So it make sens to use those format as go-for-all. But I understand your feeling, even if I am not sure that we may do better. The fact that we need to handle `controller-or-model` use case make separating controller and model a tricky case. But maybe we can merge both and have a function that retun `(controller, model string, err error)` and just set Controller & model in Run. however I am not sure that the code would be more readable.

Yeah the `controller-or-model` case does complicate things a little. We may need a special case for this indeterminate form.

Other than that though, the forms I see are all determinate, those being (laying out for sake of my own thoughts):
- `juju switch $controller:$model`
- `juju switch :$model`
- `juju switch $controller:`
- `juju switch -c $controller -m $model`
- `juju switch -c $controller`
- `juju switch -m $model`
Where `$controller` and `$model` can be `-`.

(args and flags should be mutually exclusive)

First 3 cases, I think in `Init` we should parse by splitting on `:` to determine the target controller & model which are then loaded to the `switchCommand` struct and read in `Run`

Final 3 cases, we can load straight into the struct using `SetFlags`

So if an arg is provided and doesn't contain a `:`, maybe load that into a separate attribute and branch in `Run` if it's set?

> Could need a second pair of eyes when doing that stuff... Are you open to peer prog on monday ?

I'm away until Wednesday next week, but yes I will be happy to then

Reply:
I like your suggestion, but I cannot decide in my head if it would be easier to read.

We can use the pair programming session on Wednesday, it's not an urgent task. Enjoy your time off ;)